src/fswatcher.c

```diff
@@ -0,0 +1,988 @@
+ *
+ * gcc -Wall -Wextra fswatcher.c -o fswatcher -lthread
+ *
+ */
+
+// #define DEBUG
+
+#define _REENTRANT
+#include <assert.h>
+#include <ctype.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <limits.h>
+#include <fcntl.h>
+#include <strings.h>
+#include <port.h>
+#include <errno.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <thread.h>
+#include <time.h>
+
+/*
+ * On STDIN you can send:
+ *
+ * <KEY> WATCH <pathname> [timestamp]\n
+ * <KEY> UNWATCH <pathname>\n
+ *
+ * The first will cause <pathname> to be added to the watch list. The second
+ * will cause the watch for the specified path to be removed. The <KEY> must
+ * be an integer in the range 1-4294967295 (inclusive). Leading 0's will be
+ * removed. NOTE: 0 is a special key in that it will be used in output for
+ * errors which were not directly the result of a command.
+ *
+ * When using a WATCH command with a <timestamp> argument, the timestamp must
+ * be an integer number of nanoseconds since Jan 1, 1970 00:00:00 UTC. This
+ * usually will be the time that the last event or modification was seen for
+ * this file and the data can be pulled from stat(2)'s st_mtim structure.
+ *
+ * On STDOUT you will see JSON messages that look like the following but are
+ * on a single line:
+ *
+ *  {
+ *     "changes": [array],
+ *     "code": <number>,
+ *     "is_final": true|false,
+ *     "key": <number>,
+ *     "message": "human readable string",
+ *     "pathname": "/path/which/had/event",
+ *     "result": "SUCCESS|FAIL",
+ *     "timestamp": <number>,
+ *     "type": <type>
+ *  }
+ *
+ * Where:
+ *
+ *   changes   is an array of strings indicating which changes occurred
+ *   code      is a positive integer code for an error
+ *   is_final  true when the event being printed is the last without re-watch
+ *   key       is the <KEY> for which a response corresponds
+ *   message   is a human-readable string describing response
+ *   pathname  is the path to which an event applies
+ *   result    indicates whether a command was a SUCCESS or FAILURE
+ *   timestamp timestamp in ns since 1970-01-01 00:00:00 UTC
+ *   type      is one of: event, response, error
+ *
+ * And:
+ *
+ *   "changes" is included only when (type == event)
+ *   "code" is included only when (type == error)
+ *   "is_final" is included when (type == event)
+ *   "key" is included whenever (type == response)
+ *   "message" is included whenever (type == error)
+ *   "pathname" is included whenever (type == event)
+ *   "result" is included when (type == response) value: "SUCCESS" or "FAILURE"
+ *   "timestamp" is included when (type == event)
+ *   "type" is always included
+ *
+ * Current values for "code" are in the ErrorCodes enum below.
+ *
+ * EXIT STATUS
+ *
+ *   Under normal operation, fswatcher will run until STDIN is closed or a fatal
+ *   error occurs. STDIN closing will result in exit code 0. Any other exit code
```

Comment:
The wording here's a bit strange. Maybe something like:

```If a fatal error occurs, this will result in a message of type "error" ...```

instead of:

```Any other exit code...``` ?